File: ChaosCreatures/ChaosCreatures/Models/GameState.swift
// GameState.swift
// Chaos Creatures
// Client-side game state projection received from server.
// Mirrors: packages/game-server/src/types/messages.ts ClientGameState

import Foundation

// MARK: - Stabilizer Type Enum

enum StabilizerTypeEnum: String, Codable {
    case order = "ORDER"
    case chaos = "CHAOS"
    case hybrid = "HYBRID"
}

// MARK: - Battle Stabilizer Data (in stability zone)

struct BattleStabilizerData: Codable, Identifiable {
    let instanceId: String
    let templateId: String
    let name: String
    let artUrl: String?
    let stabilizerType: StabilizerTypeEnum
    /// The "type" key from the activated_effect JSONB, decoded for display purposes.
    let activatedEffectType: String
    let isOnCooldown: Bool
    let zoneIndex: Int

    var id: String { instanceId }

    enum CodingKeys: String, CodingKey {
        case instanceId = "instance_id"
        case templateId = "template_id"
        case name
        case artUrl = "art_url"
        case stabilizerType = "stabilizer_type"
        case activatedEffect = "activated_effect"
        case isOnCooldown = "is_on_cooldown"
        case zoneIndex = "zone_index"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        instanceId = try container.decode(String.self, forKey: .instanceId)
        templateId = try container.decode(String.self, forKey: .templateId)
        name = try container.decode(String.self, forKey: .name)
        artUrl = try container.decodeIfPresent(String.self, forKey: .artUrl)
        stabilizerType = try container.decode(StabilizerTypeEnum.self, forKey: .stabilizerType)
        isOnCooldown = try container.decode(Bool.self, forKey: .isOnCooldown)
        zoneIndex = try container.decode(Int.self, forKey: .zoneIndex)
        // Decode only the "type" key from the activated_effect JSONB dict
        if let effectDict = try? container.decode([String: String].self, forKey: .activatedEffect),
           let effectType = effectDict["type"] {
            activatedEffectType = effectType
        } else {
            activatedEffectType = "UNKNOWN"
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(instanceId, forKey: .instanceId)
        try container.encode(templateId, forKey: .templateId)
        try container.encode(name, forKey: .name)
        try container.encodeIfPresent(artUrl, forKey: .artUrl)
        try container.encode(stabilizerType, forKey: .stabilizerType)
        try container.encode(isOnCooldown, forKey: .isOnCooldown)
        try container.encode(zoneIndex, forKey: .zoneIndex)
        try container.encode(["type": activatedEffectType], forKey: .activatedEffect)
    }
}

// MARK: - Client Game State

struct ClientGameState: Codable {
    let matchId: String
    let currentTurn: Int
    let activePlayer: PlayerSide
    let phase: TurnPhase
    let firstPlayer: PlayerSide
    let declaredAttackers: [String]
    let blockerAssignments: [ClientBlockerAssignment]
    let lastRollValue: Int?
    let lastRollEvent: EventType?
    let lastRollEventId: String?
    let turnTimerStarted: String?
    let turnTimerSeconds: Int
    let mySide: PlayerSide
    let myHand: [BattleCardData]
    let myDeckCount: Int
    let opponent: ClientBattlePlayer
    let me: ClientBattlePlayer
    let winner: PlayerSide?

    enum CodingKeys: String, CodingKey {
        case matchId = "match_id"
        case currentTurn = "current_turn"
        case activePlayer = "active_player"
        case phase
        case firstPlayer = "first_player"
        case declaredAttackers = "declared_attackers"
        case blockerAssignments = "blocker_assignments"
        case lastRollValue = "last_roll_value"
        case lastRollEvent = "last_roll_event"
        case lastRollEventId = "last_roll_event_id"
        case turnTimerStarted = "turn_timer_started"
        case turnTimerSeconds = "turn_timer_seconds"
        case mySide = "my_side"
        case myHand = "my_hand"
        case myDeckCount = "my_deck_count"
        case opponent, me, winner
    }

    /// Whether it is the local player's turn
    var isMyTurn: Bool {
        activePlayer == mySide
    }
}

// MARK: - Client Battle Player

struct ClientBattlePlayer: Codable {
    let playerId: String
    let side: PlayerSide
    let avatarId: String
    let currentHp: Int
    let maxHp: Int
    let currentMana: Int
    let manaCap: Int
    let instability: Int
    let board: [BattleCreatureData?]
    let handCount: Int
    let deckCount: Int
    let graveyardCount: Int
    let hasChaosSpark: Bool
    let lastEventType: EventType?
    let consecutiveMissedTurns: Int
    let isConnected: Bool
    let stabilityZone: [BattleStabilizerData]
    let stabilizersPlayedThisTurn: Int

    enum CodingKeys: String, CodingKey {
        case playerId = "player_id"
        case side
        case avatarId = "avatar_id"
        case currentHp = "current_hp"
        case maxHp = "max_hp"
        case currentMana = "current_mana"
        case manaCap = "mana_cap"
        case instability, board
        case handCount = "hand_count"
        case deckCount = "deck_count"
        case graveyardCount = "graveyard_count"
        case hasChaosSpark = "has_chaos_spark"
        case lastEventType = "last_event_type"
        case consecutiveMissedTurns = "consecutive_missed_turns"
        case isConnected = "is_connected"
        case stabilityZone = "stability_zone"
        case stabilizersPlayedThisTurn = "stabilizers_played_this_turn"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        playerId = try container.decode(String.self, forKey: .playerId)
        side = try container.decode(PlayerSide.self, forKey: .side)
        avatarId = try container.decode(String.self, forKey: .avatarId)
        currentHp = try container.decode(Int.self, forKey: .currentHp)
        maxHp = try container.decode(Int.self, forKey: .maxHp)
        currentMana = try container.decode(Int.self, forKey: .currentMana)
        manaCap = try container.decode(Int.self, forKey: .manaCap)
        instability = try container.decode(Int.self, forKey: .instability)
        board = try container.decode([BattleCreatureData?].self, forKey: .board)
        handCount = try container.decode(Int.self, forKey: .handCount)
        deckCount = try container.decode(Int.self, forKey: .deckCount)
        graveyardCount = try container.decode(Int.self, forKey: .graveyardCount)
        hasChaosSpark = try container.decode(Bool.self, forKey: .hasChaosSpark)
        lastEventType = try container.decodeIfPresent(EventType.self, forKey: .lastEventType)
        consecutiveMissedTurns = try container.decode(Int.self, forKey: .consecutiveMissedTurns)
        isConnected = try container.decode(Bool.self, forKey: .isConnected)
        stabilityZone = try container.decodeIfPresent([BattleStabilizerData].self, forKey: .stabilityZone) ?? []
        stabilizersPlayedThisTurn = try container.decodeIfPresent(Int.self, forKey: .stabilizersPlayedThisTurn) ?? 0
    }

    /// Get creature at a specific board slot (0-4)
    func creatureAt(slot: Int) -> BattleCreatureData? {
        guard slot >= 0 && slot < board.count else { return nil }
        return board[slot]
    }

    /// All non-nil creatures currently on board
    var activeCreatures: [BattleCreatureData] {
        board.compactMap { $0 }
    }

    /// Count of creatures on board
    var creatureCount: Int {
        activeCreatures.count
    }

    /// First empty slot index, or nil if board is full
    var firstEmptySlot: Int? {
        board.firstIndex(where: { $0 == nil })
    }

    /// Whether the board has any empty slots
    var hasEmptySlot: Bool {
        firstEmptySlot != nil
    }
}

// MARK: - Client Blocker Assignment

struct ClientBlockerAssignment: Codable {
    let blockerCreatureId: String
    let attackerCreatureId: String

    enum CodingKeys: String, CodingKey {
        case blockerCreatureId = "blocker_creature_id"
        case attackerCreatureId = "attacker_creature_id"
    }
}

File: packages/game-server/src/types/game-state.ts
// Chaos Creatures Game Server — Full Game State Types
// Source of truth: docs/design/02-card-data-model.md Section 13

import type {
  CardType,
  Keyword,
  EventType,
  PlayerSide,
  TurnPhase,
  FactionMechanic,
  ModifierPoolType,
  EffectType,
  TargetType,
  Duration,
  TriggerType,
  ConditionType,
  LogEntryType,
  GameMode,
  EndReason,
  SeasonRank,
  FactionId,
} from './enums';

// ─── Effect Schema (atomic game logic unit) ─────────────

export interface Condition {
  type: ConditionType;
  value?: number;
  keyword?: Keyword;
  event_type?: EventType;
}

export interface Effect {
  effect_type: EffectType;
  target: TargetType;
  value?: number;
  keyword?: Keyword;
  duration?: Duration;
  secondary_effect?: Effect;
  condition?: Condition;
}

// ─── Battle Modifier (runtime state) ─────────────

export interface BattleModifier {
  definition_id: string;
  name: string;
  pool_type: ModifierPoolType;
  faction_mechanic?: FactionMechanic;
  attunement: EventType;
  base_effect: Effect;
  attuned_effect: Effect;
  has_penalty: boolean;
  penalty_effect?: Effect;
  grants_keyword?: Keyword;
  keyword_is_attuned: boolean;
  instability_adjustment: number;
  instability_is_attuned: boolean;

  // Current state
  is_attuned_active: boolean;
  is_penalty_active: boolean;
}

// ─── Triggered Ability ─────────────

export interface TriggeredAbility {
  id: string;
  card_instance_id: string;
  evolution_step: number;
  trigger: TriggerType;
  effect: Effect;
  description: string;
  name: string;
}

// ─── Temporary Buff ─────────────

export interface TempBuff {
  effect: Effect;
  expires_at: 'END_OF_TURN' | 'UNTIL_NEXT_ROLL';
  source: string;
}

// ─── Battle Card (base for hand/deck/graveyard) ─────────────

export interface BattleCard {
  instance_id: string;
  template_id: string;
  card_type: CardType;
  name: string;
  mana_cost: number;
  art_url: string;
  base_attack?: number;
  base_health?: number;
  base_instability: number;
  innate_keywords: Keyword[];
  modifiers: BattleModifier[];
  triggered_abilities: TriggeredAbility[];
  faction_id: string;
}

// ─── Battle Creature (on board — extends BattleCard) ─────────────

export interface BattleCreature extends BattleCard {
  // Current combat stats (includes all active modifiers and buffs)
  attack: number;
  health: number;
  max_health: number;

  // State
  has_attacked: boolean;
  is_alive: boolean;

  // Instability
  instability_value: number;

  // Effective keywords right now
  active_keywords: Keyword[];
  shield_active: boolean;

  // Temporary effects
  temp_buffs: TempBuff[];

  // Board position
  board_slot: number;
}

// ─── Battle Ruin (Planar Ruin on board) ─────────────

export interface BattleRuin extends BattleCard {
  // Current stats
  health: number;
  max_health: number;

  // State
  is_alive: boolean;

  // Board position
  board_slot: number;

  // Ruin-specific effects
  passive_effect: Effect;
  destruction_penalty: Effect;

  // Ward protection (cannot be targeted by opponent modifier effects for 1 turn)
  ward_active: boolean;
}

// ─── Blocker Assignment ─────────────

export interface BlockerAssignment {
  blocker_creature_id: string;
  attacker_creature_id: string;
}

// ─── Game Log Entry ─────────────

export interface GameLogEntry {
  turn: number;
  phase: TurnPhase;
  timestamp: string;
  entry_type: LogEntryType;
  description: string;
  data: Record<string, unknown>;
}

// ─── Battle Player ─────────────

export interface BattlePlayer {
  player_id: string;
  side: PlayerSide;
  avatar_id: string;
  avatar_instability_modifier: number;

  // Health
  current_hp: number;
  max_hp: number;

  // Mana
  current_mana: number;
  mana_cap: number;

  // Instability (computed)
  instability: number;

  // Board: 5 slots, each null (empty) or occupied by a creature or ruin
  board: (BattleCreature | BattleRuin | null)[];

  // Whether a Planar Ruin is currently on the board
  ruin_on_board: boolean;

  // Hand
  hand: BattleCard[];

  // Deck (order matters — index 0 = next draw)
  deck: BattleCard[];

  // Graveyard
  graveyard: BattleCard[];

  // Chaos Spark (P2 only)
  has_chaos_spark: boolean;

  // Attunement tracking
  last_event_type: EventType | null;

  // Disconnect tracking
  consecutive_missed_turns: number;
  is_connected: boolean;

  // Deck info for match record
  deck_id: string;
  faction_id: string;
  season_rank: SeasonRank;
}

// ─── Game State (full runtime state) ─────────────

export interface GameState {
  match_id: string;
  mode: GameMode;
  started_at: string;
  current_turn: number;
  active_player: PlayerSide;
  phase: TurnPhase;

  // Players
  player_1: BattlePlayer;
  player_2: BattlePlayer;

  // First/second tracking
  first_player: PlayerSide;

  // Combat state
  declared_attackers: string[];
  blocker_assignments: BlockerAssignment[];

  // Chaos roll state
  last_roll_value: number | null;
  last_roll_event: EventType | null;
  last_roll_event_id: string | null;

  // Pending event awaiting player choice (O2 Planar Ward, O5 Fortify)
  pending_event_id?: string;

  // Turn timer
  turn_timer_started: string | null;
  turn_timer_seconds: number;

  // Game log
  log: GameLogEntry[];

  // Winner (set when game ends)
  winner: PlayerSide | null;

  // Seeded RNG seed
  rng_seed: number;
  rng_counter: number;
}

// ─── Event Definition ─────────────

export interface EventDefinition {
  id: string;
  name: string;
  event_type: EventType;
  effect: Effect;
  description: string;
  design_notes: string;
  can_backfire: boolean;
}

// ─── Combat Result Types ─────────────

export interface CombatPairResult {
  attacker_id: string;
  blocker_id: string;
  attacker_damage_dealt: number;
  blocker_damage_dealt: number;
  attacker_died: boolean;
  blocker_died: boolean;
  piercing_damage: number;
  attacker_lifesteal: number;
  blocker_lifesteal: number;
  attacker_shield_broke: boolean;
  blocker_shield_broke: boolean;
}

export interface UnblockedResult {
  attacker_id: string;
  face_damage: number;
  lifesteal: number;
}

export interface CombatResult {
  pairs: CombatPairResult[];
  unblocked: UnblockedResult[];
  deaths: Array<{ creature_id: string; side: PlayerSide; board_slot: number }>;
}

// ─── Chaos Roll Result ─────────────

export interface ChaosRollResult {
  roll: number;
  result: 'ORDER' | 'CHAOS' | 'NOTHING';
  instability: number;
}

// ─── Event Resolution Result ─────────────

export interface EventResolutionResult {
  event: EventDefinition;
  event_effect_results: EffectResult[];
  trigger_results: TriggerResult[];
}

export interface TriggerResult {
  creature_id: string;
  ability_name: string;
  effect_results: EffectResult[];
}

export interface EffectResult {
  effect_type: EffectType;
  target_ids: string[];
  value?: number;
  description: string;
}

// ─── Card Play Record (for match record) ─────────────

export interface CardPlayRecord {
  card_instance_id: string;
  player_id: string;
  turns_on_board: number;
  damage_dealt: number;
  damage_taken: number;
  triggers_fired: number;
  was_destroyed: boolean;
}

// ─── Match Record ─────────────

export interface MatchRecord {
  id: string;
  mode: GameMode;
  player_1_id: string;
  player_2_id: string;
  winner_id: string | null;
  loser_id: string | null;
  player_1_deck_id: string;
  player_2_deck_id: string;
  player_1_avatar_id: string;
  player_2_avatar_id: string;
  player_1_faction_id: string;
  player_2_faction_id: string;
  end_reason: EndReason;
  total_turns: number;
  duration_seconds: number;
  player_1_final_hp: number;
  player_2_final_hp: number;
  player_1_rank: SeasonRank;
  player_2_rank: SeasonRank;
  cards_played: CardPlayRecord[];
  total_rolls: number;
  order_events_p1: number;
  chaos_events_p1: number;
  order_events_p2: number;
  chaos_events_p2: number;
  full_log: GameLogEntry[];
  started_at: string;
  ended_at: string;
  season_id: string;
}
